### PR TITLE
Editable substruct library

### PR DESCRIPTION
**Pull request by jones-gareth**
#### Extensions to SubstructLibrary

I have a project to build an Oracle data cartridge using rdkit (should be open source).  SubstructLibrary would seem like a good solution to manage an external index.

Unfortunately the append only array model is not sufficient as I need to be able to add, retrieve and delete by a key.  I've also added a hit list queue so a consumer can get substructure matches back while the search is still running.

It would be great if this were part of rdkit- maintenance would be a lot simpler.  I expect to be adding more functionality and tests over the next few months.

#### What does this implement/fix?

* EditableSubstructureLibrary template class for managing an indexed library which contains index, mol and fingerprint holders.
* ChunkedHitlist template for producer consumer queue in searching.
* Concrete class for string key, trusted smiles molecules and pattern fingerprints
* Swig wrapper
* C++ test class
* Java test class

#### Any other comments?

* I've used boost threads for enforcing multiple read/single write in the library.  This means the boost thread and system libraries are requirements.  I added those using the cmake RDK_BUILD_THREADSAFE_SSS variable, but I don't know if that would be the best thing.
* Java test file runs fine standalone, but I can't run the test suite.  Cmake is looking for file Code/JavaWrappers/gmwrapper/runUnitTests.sh, but that is not present in the repo.
* Since this weekend I can only build the shared java library.  No big deal, but the static library is the cmake default. The issue seems to be the inclusion of the zlib library.  The swig configuration tries to pull in  /usr/lib/x86_64-linux-gnu/libz.so_static as you add _static to all required libraries, but this only works for the rdkit libraries (have I made sense, not sure I understand it myself!)
* I haven't commented the code yet.  Thought I'd wait to see if you want it.

Thanks,
Gareth

### CODE REVIEW

**Comment by bp-kelley** [comment]
Here are a few comments in no particular order.

I like this contribution and have noted the need for an index previously.

The SubstructLibrary class is a quite thin wrapper around the holders.  The holders can actually be shared between different libraries.  The holders themselves are thin wrappers around std::vectors.

At first blush I would prefer to add the indices and chunked hit list (this is great ) to the default substructlibrary and then perhaps have external filter functions.

It looks like you have added some convenience functions for adding fps from strings and the like, if we added an index holder, these could also be pushed to the core.

As for filtering, could we do these in free functions?  I can imagine two scenarios:

```
SubstructLibrary slib(...);
remove_indices(slib, ids);
SubstructLibrary slib2 = filter_remove_indices(slib, ids);
```

Alternatively, one of the design patterns that didn't make it into place was:

```
SubstructLibrary matched = slib.filter( query1,  ... );
SubstructLibrary matched2 = matched.filter( query2, ... ):
```

and by adding an index, could be:

```
for ( auto & index : slib.filter(query1).filter(query2).indices() ) {
}

```

Which I think could be quite powerful as well. 

My naming here, isn't that great, but am I making any sense?

tldr;

 (1) moving the bulk of the code to substructlibrary makes sense and will be easier to maintain
 (2) Holders were designed to be able to quickly create new holders from indices (this wasn't implemented) would it make sense to exploit this?
 (3) the chunked hit list is cool, let's add it to the core

File: Code/GraphMol/SubstructLibrary/EditableSubstructLibrary.h [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+#include <atomic>
+#include <GraphMol/Substruct/SubstructMatch.h>
+#include "ChunkedHitlist.h"
+
+#include <chrono>
+#include <thread>
+
+namespace RDKit {
+
+template <typename Key, typename MolHolder, typename FpHolder>
```
Comment (bp-kelley) [inline Code/GraphMol/SubstructLibrary/EditableSubstructLibrary.h]: I'm a little leary of the Key typename here, it makes it hard to serialize.

What are the legitimate types this could be?  Could we use an rdany?

Comment (jones-gareth) [reply Code/GraphMol/SubstructLibrary/EditableSubstructLibrary.h]: For the Key typename I just need std::string.  I templated it because Matt mentioned that he might like to have an int based index, but personally I don't see why you would need anything other than a string as (from the user's view) the index is not ordered. 

However, if we brought the index into the core library I think you would need an int key for backward compatibility with existing code as (because the underlying std::vector molecule list is scrambled on molecule deletion) for the new indexed mol holders getMol(unsigned int vector_position) would need to be getMol(Key index_value).

It would be nice to have the index in the mol holder.  I started off doing that, but changed track  because of the memory overhead of the index (I use a boost bimap for that) and because I felt I was making too many changes to core functionality.  

As first glance I thought that the external filter functions would be difficult because of the locks for shared read/exclusive write, but I guess that there's no reason not have an external function lock (it feels weird though).  I do like the idea of a substructure search filter creating a new mol holder- one issue with having an editable library is that you could end up with stale hitlists containing index values that are no longer present in the library.

What would be the best way to proceed? I could move the index stuff to MolHolderBase then implement deletions and the chunked hitlist search as external functions, but that would break the serialization stuff.

I noticed that CI failed for these changes. The Linux run ran out of time and I couldn't tell from the logs why the Xcode run failed.  Any tips for either of these would be great.

Comment (bp-kelley) [reply Code/GraphMol/SubstructLibrary/EditableSubstructLibrary.h]: @jones-gareth What do you think about this API?

```
  //! Returns the RDValue for key at the given index (assuming the library is indexed with keys)
  /*!
    \param idx       Index of the molecule in the library
  */  
  RDValue getKey(unsigned int idx) const {
    if (!keyholder.get()) throw ValueErrorException("Substruct Library made without keys")
    return keyholder->getKey(idx);
  }

  //! Returns the string value for the key at the given index (assuming the library is indexed with keys)
  //!  If the key can't be converted to a string, a ValueError is thrown
  /*!
    \param idx       Index of the molecule in the library
  */    
  std::string getKeyString(unsigned int idx) const {
    if (!keyholder.get()) throw ValueErrorException("Substruct Library made without keys")
    return keyholder->getKeyString(idx);
  }
  

  //! Return a new substruct library with only the specified indices
  SubstructLibrary filter(const std::vector<unsigned int> &indices);
  //! Return a new substructlibrary with the specified indices removed
  SubstructLibrary filter_remove(const std::vector<unsigned int> &indices);
```

The keys are optional, so overhead is opt-in.

Comment (jones-gareth) [reply Code/GraphMol/SubstructLibrary/EditableSubstructLibrary.h]: @bp-kelley Here are some ideas- let me know what you think

I like the key stuff. We'd also need to be able to retrieve molecules by key. For the filters on molholders that contain indexes we'd need functions like:

//! Return a new substruct library with only the specified keys
SubstructLibrary filter(const std::vector<RDValue> & keys)

As when you interact with the keyed library the indexes are modified on edit, but the keys may still be used to access the same molecules.

Maybe we can avoid function duplication by having a single function that expects storage array indices for molholders without indexes and index keys for molholders with keys.  

For larger libraries we (or I anyway!) would want to do the filtering in place and not create new molholders.  In the editable library I delete an item by replacing that item by the last item in the storage array and resizing the storage array by one.  I'd like to keep that efficient behavior for when we don't need to keep the old library.

//! Return a new substruct library with only the specified keys or storage indices
//! If  the library has a index keysOrIndices will be the index keys to filter by
//! If the library has on index RDValue will be of type unsigned int and keysOrIndices will be the
//! storage indices to filter
//! Set inplace = true to permanently remove the matching items from the current molholder.  The 
//! returned library will share molecules with the input library
SubstructLibrary SubstructLibrary::filter(const std::vector<RDValue> & keysOrIndices, const bool inPlace = false);

Now I've written this out, I don't like creating a new library for in place edits. That's going to be confusing in a multi-threaded environment where multiple threads are making edits.  It won't really matter as long as substruct library is just a container for molecules and fingerprints, but could be limiting going forward.

However, I liked your idea of external functions.  With an external function we could return the input library for in place operations:

//! Same as before but return the inputLibrary when inPlace == true
shared_ptr<SubstructLibrary> filter(const shared_ptr<SubstructLibrary> & inputLibrary, const std::vector<RDValue> & keysOrIndices, const bool inPlace = false);

And the same for filter_remove.

Comment (bp-kelley) [reply Code/GraphMol/SubstructLibrary/EditableSubstructLibrary.h]: In the implementation I'm playing with we could do both.  Create a new library or make the existing.

I also didn't quite realize that you needed thread-safe editing for adding as well as removing.  This makes sense.  So to clarify, your use cases are:

  (1) multithreaded manipulation of the substructure contents
  (2) all threads share the same library
  (3) when doing a substructure run, no one should add or remove, but other threads should be able to search  (RWLock?) or do we not care here.
  (4) map keys -> indices and indices to keys.

Keys can be string or int.

And finally, you have implemented a "return as soon as you get something" search implementation.
 
I think we can accomplish this with another Data Holder, and it might be nice for it to store more than indices.  Every holder should derive from a base class that allows one to remove an index or create a clone of itself.  This should (maybe!) satisfy your use case and mine which is:

```
 hits = match(query).match(query2).match(query3);
```

If we can't get this to work out, we'll just use your implementation :)

Comment (jones-gareth) [reply Code/GraphMol/SubstructLibrary/EditableSubstructLibrary.h]: I guess maybe we don't need to lock on addition (though I do) since we are just appending to the vector storage rather exchanging.  In the current application I use boost::shared_mutex for exclusive write, shared read/search (don't think that std::shared_mutex is ready yet).

I think your way will work as long as the library just contains the holder- for in place editing we can create new libraries but share the underlying mol holder.  I use the molholder (and fp holder) as the unique collection of molecules and you get the fluent programming style through the library.

So my use cases are:

(1) multithreaded manipulation of the substructure contents
(2) all threads share the same dataholder(s) containing molholder/fpholder/index
(3) when doing a substructure run, no one should add or remove, but multiple other threads should be able to search (boost::shared_lock<boost::shared_mutex>).
(4) map keys -> indices and indices to keys.
(5) SubstructLibrary is a lightweight container for the dataholder(s).
(6) Only a single thread can edit and no searching can be done during editing (boost::unique_lock<boost::shared_mutex>)

For your case (chaining queries) I think it is fine to always return new compounds- I can't imagine a use case where you would replace a set of compounds with a common substructure (though we could do that).

For edit  I do:
```
outputLibrary = inputLibrary.filter(keys, inPlace=true).add(keys, mols, inPlace=True)
```
Where inputLibrary and outputLibrary have the same underlying holder.
Any you could do
```
outputLibrary = inputLibrary.filter(keys, inPlace=false).add(keys, mols, inPlace=false)
``
In this case inputLibrary would retain the original compounds and output library would have it's own set.

If you agree I think we're almost there!

**Comment by bp-kelley** [comment]
@jones-gareth I'll have a straw man example for you to look at in a day or so.  Perhaps we should close this PR and work on one from my fork?  We can always reopen this is necessary.

**Comment by jonesgareth** [comment]
Sounds good.  Fine by me if you want close and work in your branch.  I am visiting family in Scotland right now and probably won't be able to do much on it this week.

**Comment by bp-kelley** [comment]
@jones-gareth I'm back on this one, expect something this week after I get everything sorted again.

Super apologies for the delay.

**Comment by jones-gareth** [comment]
@bp-kelley Hi Brain- this is not currently something that I would use (it's just easier to do all the multi-threading stuff in Java), so no need to rush.  I'm happy to continue work on it if it is useful to you and others.

**Comment by bp-kelley** [comment]
I’m taking a measured approach anyway, first adding keys then adding editable support.

I’m curious if you are making a complete replacement or using the substructlibrary itself.  I expect the former as otherwise java doesn’t know what threads are running.

----
Brian Kelley

**Comment by jones-gareth** [comment]
@bp-kelley Yes, I ended up just coding everything in Java.  I had other requirements that would have meant larger changes to the C code base- for example I had to write code to accommodate rubbish smiles by trying things with and without sanitization. Sorry for the late reply, I was on vacation.
